pkg/retry/azure_error.go

```diff
@@ -327,47 +327,85 @@ func HasStatusForbiddenOrIgnoredError(err error) bool {
 	}
 
-	return rawErrorMap["error"], nil
+	return matches[1], matches[2], nil
 }
 
-// IsErrorLoadBalancerInUseByVirtualMachineScaleSet determines if the Error is
+// isErrorLoadBalancerInUseByVirtualMachineScaleSet determines if the Error is
 // LoadBalancerInUseByVirtualMachineScaleSet
-func IsErrorLoadBalancerInUseByVirtualMachineScaleSet(rawError string) bool {
+func isErrorLoadBalancerInUseByVirtualMachineScaleSet(rawError string) bool {
 	return strings.Contains(rawError, "LoadBalancerInUseByVirtualMachineScaleSet")
 }
 
-// GetVMSSMetadataByRawError gets the vmss name by parsing the error message
-func GetVMSSMetadataByRawError(rawError string) (string, string, error) {
-	if !IsErrorLoadBalancerInUseByVirtualMachineScaleSet(rawError) {
-		return "", "", nil
+const (
+	// OperationNotAllowed is an umbrella errrfor a lot of errors
+	OperationNotAllowed string = "OperationNotAllowed"
+	// QuotaExceeded falls under OperationNotAllowed error code but we make it more specific here
+	QuotaExceeded string = "QuotaExceeded"
+)
+
+// ServiceRawError wraps the RawError field satisfying autorest.ServiceError
+type ServiceRawError struct {
+	ServiceError *azure.ServiceError `json:"error,omitempty"`
+}
+
+// ServiceErrorMessage returns the message associated with the autorest.ServiceError body
+func (err *Error) ServiceErrorMessage() string {
+	if err == nil || err.RawError == nil {
+		return ""
 	}
 
-	rawErrorInfo, err := ParseRawError(rawError)
-	if err != nil {
-		klog.Warningf("GetVMSSMetadataByRawError: failed to parse raw error: %v", err)
-		return "", "", nil
+	sre := ServiceRawError{}
+	marshalErr := json.Unmarshal([]byte(err.RawError.Error()), &sre)
+	if marshalErr != nil {
+		return ""
+	}
+	if sre.ServiceError == nil {
+		return ""
 	}
+	return sre.ServiceError.Message
+}
 
-	reg := regexp.MustCompile(`.*/subscriptions/(?:.*)/resourceGroups/(.*)/providers/Microsoft.Compute/virtualMachineScaleSets/(.+).`)
-	matches := reg.FindStringSubmatch(rawErrorInfo.Message)
-	if len(matches) != 3 {
-		return "", "", fmt.Errorf("GetVMSSMetadataByRawError: couldn't find a VMSS resource Id from error message %s", rawErrorInfo.Message)
+// ServiceErrorCode returns the code associated with the autorest.ServiceError body
+func (err *Error) ServiceErrorCode() string {
+	if err == nil || err.RawError == nil {
+		return ""
 	}
 
-	return matches[1], matches[2], nil
+	sre := ServiceRawError{}
+	marshalErr := json.Unmarshal([]byte(err.RawError.Error()), &sre)
+	if marshalErr != nil {
+		return ""
+	}
+	if sre.ServiceError == nil {
+		return ""
+	}
+	return classifyErrorCode(*sre.ServiceError)
+}
+
+func classifyErrorCode(sre azure.ServiceError) string {
+	if sre.Code == OperationNotAllowed {
+		return getOperationNotAllowedReason(sre.Message)
+	}
+	return sre.Code
+}
+
+// getOperationNotAllowedReason attempts to better classify OperationNotAllowed errors
+// by looking at the message
+func getOperationNotAllowedReason(msg string) string {
```

Comment:
This does a further classification on "OperationNotAllowed" errors to detect if the error is a quota one.